backend/push.js

```diff
@@ -1,53 +1,294 @@
+    socket.subscriptions?.forEach(channelID => {
+      server.pushSubscriptions[subscriptionId].subscriptions.add(channelID)
+    })
+    saveSubscription(server, subscriptionId)
+    console.error('@@@@ADD PUSH SUBSCRIPTION', subscriptionId, subscription.endpoint)
   },
-  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] (payload) {
+  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] () {
     const socket = this
-    const subscriptionId = JSON.parse(payload)
+    const { server, pushSubscriptionId: subscriptionId } = socket
 
-    delete socket.server.pushSubscriptions[subscriptionId]
-  },
-  [PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION]: async function (payload) {
-    const pushSubscriptions = this.server.pushSubscriptions
-    const data = JSON.parse(payload)
-    const sendPush = (sub) => pushController.sendNotification(
-      sub, JSON.stringify({ title: data.title, body: data.body })
-    )
+    if (subscriptionId) {
+      removeSubscription(server, subscriptionId)
+    }
+  }
+}
 
-    // NOTE: if the payload contains 'endpoint' field, send push-notification to that particular subscription.
-    //       otherwise, iterate all existing subscriptions and broadcast the push-notification to all.
+const removeSubscription = (server, subscriptionId) => {
+  const subscription = server.pushSubscriptions[subscriptionId]
+  delete server.pushSubscriptions[subscriptionId]
+  if (server.subscribersByChannelID) {
+    subscription.subscriptions.forEach((channelID) => {
+      server.subscribersByChannelID[channelID].delete(subscription)
+    })
+  }
+  deleteSubscriptionFromIndex(subscriptionId).then(() => {
+    return sbp('chelonia/db/delete', `_private_webpush_${subscriptionId}`)
+  }).catch((e) => console.error(e, 'Error removing subscription', subscriptionId))
+}
 
-    if (data.endpoint) {
-      const subscription = pushSubscriptions[data.endpoint]
+const deleteClient = (subscriptionId) => {
+  const server = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+  removeSubscription(server, subscriptionId)
+}
 
-      await sendPush(subscription)
-    } else {
-      for (const subscription of Object.values(pushSubscriptions)) {
-        await sendPush(subscription)
-      }
+const encryptPayload = async (subcription: Object, data: string) => {
+  const readableStream = new Response(data).body
+  const [asPublic, IKM] = await subcription.encryptionKeys
+
+  return encrypt(aes128gcm, readableStream, 32768, asPublic, IKM).then(async (bodyStream) => {
+    const chunks = []
+    const reader = bodyStream.getReader()
+    for (;;) {
+      const { done, value } = await reader.read()
+      if (done) break
+      chunks.push(new Uint8Array(value))
+    }
+    return Buffer.concat(chunks)
+  })
+}
+
+export const postEvent = async (subscription: Object, event: string): Promise<void> => {
+  const authorization = await vapidAuthorization(subscription.endpoint)
+  const body = event
+    ? await encryptPayload(subscription, event)
+    : undefined
```

Comment:
Under what situation would `body` be set to `undefined` and are we sure we want to still use it with `fetch` in that case?

Reply:
Originally, web push notifications didn't have a body, and then the spec was amended to support sending an (encrypted) body. Although we don't currently use these 'bodyless' notifications, they are possible.